addon/utils/notify-changes.js

```diff
@@ -0,0 +1,53 @@
+import Ember from 'ember';
+import { notifyPropertyChange as _notifyPropertyChange } from '@ember/object';
+import { assert } from '@ember/debug';
+import { DEBUG } from '@glimmer/env';
+
+const { changeProperties } = Ember;
+
+const { propertyDidChange } = Ember;
+export let notifyPropertyChange;
+
+const HasNotifyPropertyChange = _notifyPropertyChange !== undefined;
+if (HasNotifyPropertyChange) {
+  notifyPropertyChange = _notifyPropertyChange;
+} else {
+  notifyPropertyChange = propertyDidChange;
+}
+
+const StoreToChanges = new WeakMap();
+
+export function beginChanges(store) {
+  StoreToChanges.set(store, []);
+}
+
+export function deferPropertyChange(store, obj, key) {
+  if (!StoreToChanges.has(store)) {
+    beginChanges(store);
+  }
+  let changes = StoreToChanges.get(store);
+  changes.push(obj, key);
+}
+
+export function flushChanges(store) {
+  let changes = StoreToChanges.get(store) || [];
+  changeProperties(() => {
+    for (let i = 0; i < changes.length; i += 2) {
```

Comment:
I assume we know this data layout is best for perf?

Reply:
what's the layout you're referring to specifically?